src/coreclr/jit/rangecheckcloning.cpp

```diff
@@ -552,34 +563,69 @@ PhaseStatus Compiler::optRangeCheckCloning()
-            if ((largestGroup == nullptr) || (value->Height() > largestGroup->Height()))
+            if ((value->Height() >= MIN_CHECKS_PER_GROUP) && !DoesComplexityExceed(this, value))
             {
-                if (DoesComplexityExceed(this, value))
-                {
-                    continue;
-                }
-                largestGroup = value;
+                groups.Push(value);
             }
         }
 
-        if (largestGroup == nullptr)
+        if (groups.Height() == 0)
         {
             JITDUMP("No suitable group of bounds checks in the block - bail out.\n");
             continue;
         }
 
-        if (largestGroup->Height() < MIN_CHECKS_PER_GROUP)
+        // We have multiple groups of bounds checks in the block.
+        // let's pick a group that appears first in the block and the one whose last bounds check
+        // appears last in the block.
+        //
+        BoundsCheckInfoStack* firstGroup = groups.Top();
+        BoundsCheckInfoStack* lastGroup  = groups.Top();
+        for (int i = 0; i < groups.Height(); i++)
         {
-            JITDUMP("Not enough bounds checks in the largest group - bail out.\n");
-            continue;
+            BoundsCheckInfoStack* group      = groups.Bottom(i);
+            int                   firstStmt  = group->Bottom().stmtIdx;
+            int                   secondStmt = group->Top().stmtIdx;
+            if (firstStmt < firstGroup->Bottom().stmtIdx)
+            {
+                firstGroup = group;
+            }
+            if (secondStmt > lastGroup->Top().stmtIdx)
+            {
+                lastGroup = group;
+            }
+        }
+
+        // We're going to clone for the first group.
+        // But let's see if we can extend the end of the group so future iterations
+        // can fit more groups in the same block.
+        //
+        Statement* lastStmt = firstGroup->Top().stmt;
+
+        int firstGroupStarts = firstGroup->Bottom().stmtIdx;
+        int firstGroupEnds   = firstGroup->Top().stmtIdx;
+        int lastGroupStarts  = lastGroup->Bottom().stmtIdx;
+        int lastGroupEnds    = lastGroup->Top().stmtIdx;
+
+        // The only requirement is that both groups must overlap - we don't want to
+        // end up cloning unrelated statements between them (not a correctness issue,
+        // just a heuristic to avoid cloning too much).
+        //
+        if (firstGroupEnds < lastGroupEnds && firstGroupEnds >= lastGroupStarts)
+        {
+            lastStmt = lastGroup->Top().stmt;
         }
 
-        JITDUMP("Cloning bounds checks in " FMT_BB "\n", block->bbNum);
-        block    = optRangeCheckCloning_DoClone(this, block, largestGroup);
+        JITDUMP("Cloning bounds checks in " FMT_BB " from " FMT_STMT " to " FMT_STMT "\n", block->bbNum,
+                firstGroup->Bottom().stmt->GetID(), lastStmt->GetID());
+
+        block    = optRangeCheckCloning_DoClone(this, block, firstGroup, lastStmt);
```

Comment:
Is `block` here guaranteed to be nonnull in the fast-path case?

Might be a bit clearer to have `optRangeCheckCloning_DoClone` return the next block to visit

Reply:
Addressed